File: packages/nitro-renderer/src/room/object/visualization/data/ColorData.ts
export class ColorData {
    public static DEFAULT_COLOR: number = 0xffffff;

    private _colors: number[] = [];

    constructor(layerCount: number) {
        for (let i = 0; i < layerCount; i++) this._colors.push(ColorData.DEFAULT_COLOR);
    }

    public dispose(): void {
        this._colors = [];
    }

    public getLayerColor(layerId: number): number {
        return this._colors[layerId] ?? ColorData.DEFAULT_COLOR;
    }

    public setColorLayer(layerId: number, color: number): void {
        const existing = this._colors[layerId];

        if (!existing) return;

        this._colors[layerId] = color;
    }
}

File: packages/nitro-swift/Sources/NitroRoom/Furniture/ColorData.swift
/// Swift port of `ColorData` (packages/nitro-renderer/src/room/object/visualization/data/ColorData.ts).
public final class ColorData {
    public static let defaultColor: UInt32 = 0xFFFFFF

    private var colors: [UInt32]

    public init(layerCount: Int) {
        colors = Array(repeating: ColorData.defaultColor, count: max(0, layerCount))
    }

    public func getLayerColor(_ layerId: Int) -> UInt32 {
        guard layerId >= 0, layerId < colors.count else { return ColorData.defaultColor }

        return colors[layerId]
    }

    public func setColorLayer(_ layerId: Int, _ color: UInt32) {
        // TS only overwrites an already-initialized slot (`if (!existing) return`), which in
        // practice always exists since the array is pre-filled to layerCount in the constructor.
        guard layerId >= 0, layerId < colors.count else { return }

        colors[layerId] = color
    }
}
